psl-core/src/main/java/org/linqs/psl/model/rule/AbstractRule.java

```diff
@@ -23,45 +23,64 @@
 
 /**
  * Base class for all (first order, i.e., not ground) rules.
- *
- * Care should always be taken when creating multiple instance of the same rule (e.g. through deserialization) or checking rules for equality.
- * In general, there should only be one instance of each rule, but this is not tightly enforced.
- * Rule hashes are by default their identity hash code (object id).
- * This means that it may be possible to have rules that are equal(), but that do not have matching hashes.
  */
 public abstract class AbstractRule implements Rule {
-    private static Map<Integer, AbstractRule> rules = new HashMap<Integer, AbstractRule>();
+    private static final Map<Integer, Rule> rules = new HashMap<Integer, Rule>();
 
-    protected final String name;
+    protected String name;
+    protected int hashcode;
 
-    public static AbstractRule getRule(int hashcode) {
+    public static Rule getRule(int hashcode) {
         return rules.get(hashcode);
     }
 
-    public AbstractRule(String name) {
-        this(name, true);
-    }
+    protected AbstractRule() {}
```

Comment:
Put a comment here that warns people that use this constructor that they must initialize name and hashcode.

Also, I would either make name/hashcode final or explicitly set their empty values here.